flink-autoscaler/src/main/java/org/apache/flink/autoscaler/ScalingMetricCollector.java

```diff
@@ -181,9 +186,12 @@ private static Instant getWindowFullTime(
     }
 
     @VisibleForTesting
-    protected Instant getJobUpdateTs(JobDetailsInfo jobDetailsInfo) {
-        return Instant.ofEpochMilli(
-                jobDetailsInfo.getTimestamps().values().stream().max(Long::compare).get());
+    protected Instant getJobSwitchToRunningTs(JobDetailsInfo jobDetailsInfo) {
+        final Map<JobStatus, Long> timestamps = jobDetailsInfo.getTimestamps();
+
+        final Long runningTs = timestamps.get(JobStatus.RUNNING);
+        checkState(runningTs != null, "Unable to find when the job was switched to RUNNING.");
```

Comment:
Checking `JobStatus.RUNNING` exists here. If it doesn't exist, throw exception directly and exit the current scaling.

Reply:
Generally, this code will only ever be called when the job is in running mode, see: https://github.com/apache/flink-kubernetes-operator/blob/6be277cd0c7421f4822c49296198f4a34b2cd721/flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobAutoScalerImpl.java#L96

We can add this invariant here but it would only defend against a Byzantine behavior.

Reply:
The old way of getting the "jobUpdateTs" was correct but maybe this one will defend us in case of refactors.

Reply:
Yeah, I see this code only be called when the job is running. I updated it for 2 reasons:

1.  This job may be failed during scaling. And `jobDetailsInfo` is fetched by rest api, the latest status may be not RUNNING.
2. As you mentioned, it will more robust if we have some refactors in the future.

Reply:
>We can add this invariant here but it would only defend against a Byzantine behavior.

I don't quite agree - I find it it a bit one-sided to look at public classes and methods only from the perspective of the implicit context in which they are **currently** called. 
I believe the check is appropriate - nothing in the contract or comments  of `updateMetrics` states the scope in which it is supposed to be executed. The class is also open for extension - someone can later override methods partially without being clear about the implicit assumptions. An alternative is to seal everything and make this class only available to `JobAutoscalerImpl`.